org.eclipse.scanning.device.ui/src/org/eclipse/scanning/device/ui/model/TypeEditor.java

```diff
@@ -0,0 +1,162 @@
+package org.eclipse.scanning.device.ui.model;
+
+import java.lang.reflect.Constructor;
+
+import javax.inject.Inject;
+
+import org.eclipse.richbeans.api.binding.IBeanController;
+import org.eclipse.richbeans.api.event.ValueAdapter;
+import org.eclipse.richbeans.api.event.ValueEvent;
+import org.eclipse.richbeans.binding.BeanService;
+import org.eclipse.richbeans.widgets.internal.GridUtils;
+import org.eclipse.scanning.api.annotation.scan.AnnotationManager;
+import org.eclipse.scanning.api.annotation.ui.TypeDescriptor;
+import org.eclipse.scanning.api.device.IRunnableDeviceService;
+import org.eclipse.scanning.api.device.IScannableDeviceService;
+import org.eclipse.scanning.device.ui.Activator;
+import org.eclipse.scanning.device.ui.ServiceHolder;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.ui.services.IDisposable;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.packageadmin.PackageAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("deprecation")
+public class TypeEditor extends Composite {
+	
+	private static final Logger logger = LoggerFactory.getLogger(TypeEditor.class);
+
+	private IBeanController controller;
+	
+	public TypeEditor(Composite parent, int style) {
+		super(parent, style);
+		setLayout(new GridLayout(1, false));
+		GridUtils.removeMargins(this);
+	}
+	
+	public void setModel(Object model) throws Exception {
+		
+		deactivate();
+	
+		Object ui = createUserInterface(model);
+		controller = BeanService.getInstance().createController(ui, model);
+		controller.beanToUI();
+		controller.switchState(true);
+		controller.addValueListener(new ValueAdapter("ui change") {		
+			@Override
+			public void valueChangePerformed(ValueEvent e) {
+				// Save the values
+				try {
+					controller.uiToBean();
+				} catch (Exception e1) {
+					logger.error("Problem recording bean data!", e1);
+				}
+			}
+		});
+	}
+	
+	public Object getModel() throws Exception {
+		
+		controller.uiToBean();
+        return controller.getBean();
+	}
+	
+	private Object createUserInterface(Object model) throws Exception {

```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)